internal/global/internal_logging.go

```diff
@@ -27,37 +27,36 @@ import (
 //
 // The default logger uses stdr which is backed by the standard `log.Logger`
 // interface. This logger will only show messages at the Error Level.
-var globalLogger logr.Logger = stdr.New(log.New(os.Stderr, "", log.LstdFlags|log.Lshortfile))
-var globalLoggerLock = &sync.RWMutex{}
+var globalLogger = func() *atomic.Value {
+	dl := &atomic.Value{}
+	dl.Store(stdr.New(log.New(os.Stderr, "", log.LstdFlags|log.Lshortfile)))
+	return dl
+}()
 
 // SetLogger overrides the globalLogger with l.
 //
 // To see Info messages use a logger with `l.V(1).Enabled() == true`
 // To see Debug messages use a logger with `l.V(5).Enabled() == true`.
 func SetLogger(l logr.Logger) {
-	globalLoggerLock.Lock()
-	defer globalLoggerLock.Unlock()
-	globalLogger = l
+	globalLogger.Store(l)
```

Comment:
This will panic unless the user passes a `logr.Logger` that has a concrete type of `stdr.logger`: https://pkg.go.dev/sync/atomic#Value.Store

![20221216_101056](https://user-images.githubusercontent.com/5543599/208161970-52929eba-6c2b-4f6c-825c-2c916ff0cd40.png)

I'm guessing we want to use an atomic pointer here similar to https://github.com/open-telemetry/opentelemetry-go/pull/3543

Reply:
Or we could use the "holder" pattern we use elsewhere: https://github.com/open-telemetry/opentelemetry-go/blob/7a60bc785d669fa6ad26ba70e88151d4df631d90/sdk/metric/manual_reader.go#L101-L103

Reply:
yeah.. fixed.